src/libbson/tests/test-json.c

```diff
@@ -2841,6 +2841,701 @@ test_bson_as_json_multi_object (void)
+      BSON_MAX_LEN_UNLIMITED,
+      "{ \"v\" : { \"$timestamp\" : { \"t\" : 461394000, \"i\" : 2 } } }");
+
+   bson_destroy (b);
+}
+
+static void
+test_bson_as_json_with_opts_minkey (void)
+{
+   bson_t *b;
+
+   b = bson_new ();
+   BSON_ASSERT (BSON_APPEND_MINKEY (b, "v"));
+
+   TEST_BSON_AS_JSON_WITH_OPTS (b,
+                                BSON_JSON_MODE_RELAXED,
+                                BSON_MAX_LEN_UNLIMITED,
+                                "{ \"v\" : { \"$minKey\" : 1 } }");
+
+   TEST_BSON_AS_JSON_WITH_OPTS (b,
+                                BSON_JSON_MODE_CANONICAL,
+                                BSON_MAX_LEN_UNLIMITED,
+                                "{ \"v\" : { \"$minKey\" : 1 } }");
+
+   TEST_BSON_AS_JSON_WITH_OPTS (
+      b, BSON_JSON_MODE_RELAXED, 17, "{ \"v\" : { \"$minKe");
+
+   TEST_BSON_AS_JSON_WITH_OPTS (
+      b, BSON_JSON_MODE_CANONICAL, 17, "{ \"v\" : { \"$minKe");
+
+   bson_destroy (b);
+}
+
+static void
+test_bson_as_json_with_opts_maxkey (void)
+{
+   bson_t *b;
+
+   b = bson_new ();
+   BSON_ASSERT (BSON_APPEND_MAXKEY (b, "v"));
+
+   TEST_BSON_AS_JSON_WITH_OPTS (b,
+                                BSON_JSON_MODE_RELAXED,
+                                BSON_MAX_LEN_UNLIMITED,
+                                "{ \"v\" : { \"$maxKey\" : 1 } }");
+
+   TEST_BSON_AS_JSON_WITH_OPTS (b,
+                                BSON_JSON_MODE_CANONICAL,
+                                BSON_MAX_LEN_UNLIMITED,
+                                "{ \"v\" : { \"$maxKey\" : 1 } }");
+
+   TEST_BSON_AS_JSON_WITH_OPTS (
+      b, BSON_JSON_MODE_RELAXED, 17, "{ \"v\" : { \"$maxKe");
+
+   TEST_BSON_AS_JSON_WITH_OPTS (
+      b, BSON_JSON_MODE_CANONICAL, 17, "{ \"v\" : { \"$maxKe");
+
+   bson_destroy (b);
+}
+
+static void
+test_bson_as_json_with_opts_decimal128 (void)
+{
+   bson_t *b;
+   bson_decimal128_t dec;
+
+   dec.high = 0x3040ffffffffffffULL;
+   dec.low = 0xffffffffffffffffULL;
+
+   b = bson_new ();
+   BSON_ASSERT (BSON_APPEND_DECIMAL128 (b, "v", &dec));
+
+   TEST_BSON_AS_JSON_WITH_OPTS (b,
+                                BSON_JSON_MODE_RELAXED,
+                                BSON_MAX_LEN_UNLIMITED,
+                                "{ \"v\" : { \"$numberDecimal\" : "
+                                "\"5192296858534827628530496329220095\" } }");
+
+   TEST_BSON_AS_JSON_WITH_OPTS (b,
+                                BSON_JSON_MODE_CANONICAL,
+                                BSON_MAX_LEN_UNLIMITED,
+                                "{ \"v\" : { \"$numberDecimal\" : "
+                                "\"5192296858534827628530496329220095\" } }");
+
+   TEST_BSON_AS_JSON_WITH_OPTS (
+      b,
+      BSON_JSON_MODE_RELAXED,
+      53,
+      "{ \"v\" : { \"$numberDecimal\" : \"51922968585348276285304");
+
+   TEST_BSON_AS_JSON_WITH_OPTS (
+      b,
+      BSON_JSON_MODE_CANONICAL,
+      53,
+      "{ \"v\" : { \"$numberDecimal\" : \"51922968585348276285304");
+
+   bson_destroy (b);
+}
+
```

Comment:
I suggest one more test. Create a BSON document with every type. Serialize it with no max length limit. Then iterate `i = 0..full string length` and serialize it with a max length of `i`. Assert that the produced string is exactly `i` characters.

Here is an example document:
https://gist.github.com/kevinAlbs/0807bde6f9e5b166abafb502a37f0da5

Reply:
Oh that looks awesome! That gives me a lot of confidence for these changes.

Reply:
👍